<?xml version="1.0"?>

<launch>
  <arg name="network_files" default="[$(find neural_tracker)/saved_networks/TESTpolicies6D_PT_h40_h40.pkl]" />

  <!-- Crazyflie radio address. -->
  <arg name="uri" default="radio://0/25/2M/E7E7E7E701" />

  <!-- Topics. -->
  <arg name="position_velocity_yaw_state_topic" default="/state/position_velocity_yaw" />
  <arg name="position_velocity_state_topic" default="/state/position_velocity" />
  <arg name="reference_state_topic" default="/ref/planner" />
  <arg name="in_flight_topic" default="/in_flight" />

  <arg name="lqr_control_topic" default="/control/lqr" />
  <arg name="optimal_control_topic" default="/control/optimal" />
  <arg name="merged_control_topic" default="/control/merged" />
  <arg name="takeoff_control_topic" default="/control/takeoff" />
  <arg name="final_control_topic" default="/control/final" />

  <arg name="sensor_topic" default="/sensor" />
  <arg name="controller_id_topic" default="/ref/controller_id" />
  <arg name="traj_topic" default="/traj" />
  <arg name="traj_vis_topic" default="/vis/traj" />
  <arg name="bound_vis_topic" default="/vis/bound" />
  <arg name="known_env_vis_topic" default="/vis/known_env" />
  <arg name="true_env_vis_topic" default="/vis/true_env" />
  <arg name="sensor_radius_vis_topic" default="/vis/sensor" />
  <arg name="request_traj_topic" default="/request_traj" />
  <arg name="trigger_replan_topic" default="/replan" />

  <!-- Frames of reference. -->
  <arg name="fixed_frame" default="world" />
  <arg name="tracker_frame" default="tracker" />
  <arg name="planner_frame" default="planner" />
  <arg name="mocap_frame" default="/vicon/cf7/cf7" />

  <!-- Service names. -->
  <arg name="optimal_control_name" default="/optimal_control" />
  <arg name="tracking_bound_name" default="/tracking_bound" />
  <arg name="switching_bound_name" default="/switching_bound" />
  <arg name="switching_time_name" default="/switching_time" />
  <arg name="switching_distance_name" default="/switching_distance" />
  <arg name="priority_name" default="/priority" />
  <arg name="max_planner_speed_name" default="/max_planner_speed" />
  <arg name="best_time_name" default="/best_time" />

  <!-- State bounds (x, y, z, x_dot, y_dot, z_dot). -->
  <arg name="state_lower_bound" default="[-1.5, -1.5, 0.0, -1.0, -1.0, -1.0]" />
  <arg name="state_upper_bound" default="[2.5, 2.5, 3.0, 1.0, 1.0, 1.0]" />

  <!-- Control bounds (pitch, roll, thrust). -->
  <arg name="control_lower_bound" default="[-0.15, -0.15, 7.81]" />
  <arg name="control_upper_bound" default="[0.15, 0.15, 11.81]" />

  <!-- Goal point. -->
  <arg name="goal_x" default="1.5" />
  <arg name="goal_y" default="1.5" />
  <arg name="goal_z" default="1.8" />

  <!-- Random seed for obstacle locations -->
  <arg name="random_seed" default="19" />

  <!-- Sensor and environment params. -->
  <arg name="sensor_radius" default="1.0" />
  <arg name="num_obstacles" default="2" />
  <arg name="min_obstacle_radius" default="0.1" />
  <arg name="max_obstacle_radius" default="0.15" />

  <!-- Time steps. -->
  <arg name="simulator_dt" default="0.001" />
  <arg name="estimator_dt" default="0.005" />
  <arg name="tracker_dt" default="0.01" />
  <arg name="sensor_dt" default="0.1" />

  <!-- Control merge mode. -->
  <arg name="merger_mode" default="PRIORITIZED" />

  <!-- Meta planning params. -->
  <arg name="max_meta_runtime" default="0.75" />
  <arg name="max_meta_connection_radius" default="10.0" />

  <!-- Value function server params. -->
  <arg name="numerical_mode" default="false" />
  <arg name="num_values" default="2" />
  <arg name="value_directories"
       default="[speed_7_tenths/, speed_7_tenths_to_4_tenths/, speed_4_tenths/, speed_4_tenths_to_1_tenths/]" />
  <arg name="max_speeds" default="[0.125, 0.125]" /> <!-- HACK! -->
  <arg name="max_velocity_disturbances" default="[0.0, 0.0]" />
  <arg name="max_acceleration_disturbances" default="[0.0, 0.0]" />

  <!-- Controller params. -->
  <arg name="tracker_x_dim" default="6" />
  <arg name="lqr_x_dim" default="7" />
  <arg name="tracker_u_dim" default="3" />
  <arg name="lqr_u_dim" default="4" />

  <arg name="lqr_K_file" default="$(find crazyflie_lqr)/references/K_poor.txt" />
  <arg name="lqr_u_ref_file" default="$(find crazyflie_lqr)/references/u_ref_hover.txt" />

  <!-- Simulator params. -->
  <arg name="takeoff_hover_x" default="0.0" />
  <arg name="takeoff_hover_y" default="0.0" />
  <arg name="takeoff_hover_z" default="1.9" />

  <arg name="record" default="false" />

  <!-- Record a rosbag. -->
  <node pkg="rosbag"
	      type="record"
	      name="recorder"
	      args="-a -o $(find meta_planner)/bagfiles/sw_demo_recording.bag"
	      if="$(arg record)">
  </node>


  <!-- TF static publisher to create a base_link frame for the tracker. -->
  <node pkg="tf"
        type="static_transform_publisher"
        name="tracker_base_link_publisher"
        args="0 0 0 0 0 0 1 $(arg tracker_frame) $(arg tracker_frame)/base_link 100">
  </node>

  <node pkg="tf"
        type="static_transform_publisher"
        name="tracker_mocap_publisher"
        args="0 0 0 0 0 0 1 $(arg mocap_frame) $(arg tracker_frame) 100">
  </node>

  <!-- Crazyflie interface. -->
  <include file="$(find crazyflie_driver)/launch/crazyflie_server.launch" />

  <group ns="crazyflie">
    <include file="$(find crazyflie_driver)/launch/crazyflie_add.launch">
      <arg name="uri" value="$(arg uri)" />
      <arg name="tf_prefix" value="$(arg tracker_frame)" />
      <arg name="enable_logging" value="True" />
    </include>
  </group>

  <!-- Takeoff server node. -->
  <node name="takeoff_server"
        pkg="crazyflie_takeoff"
        type="takeoff_server_node"
        output="screen">

    <param name="topics/reference" value="$(arg reference_state_topic)" />
    <param name="topics/control" value="$(arg takeoff_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />

    <param name="hover/x" value="$(arg takeoff_hover_x)" />
    <param name="hover/y" value="$(arg takeoff_hover_y)" />
    <param name="hover/z" value="$(arg takeoff_hover_z)" />
  </node>

  <!-- Takeoff server node. -->
  <node name="takeoff_control_filter"
        pkg="crazyflie_takeoff"
        type="takeoff_control_filter_node"
        output="screen">

    <param name="topics/takeoff_control" value="$(arg takeoff_control_topic)" />
    <param name="topics/commanded_control" value="$(arg merged_control_topic)" />
    <param name="topics/final_control" value="$(arg final_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />

    <param name="hover/x" value="$(arg takeoff_hover_x)" />
    <param name="hover/y" value="$(arg takeoff_hover_y)" />
    <param name="hover/z" value="$(arg takeoff_hover_z)" />
  </node>

  <!-- Converter for custom command velocity messages. -->
  <node name="cmd_vel_converter"
        pkg="crazyflie_control_merger"
        type="cmd_vel_converter_node"
        output="screen">

    <param name="topics/control" value="$(arg final_control_topic)" />
    <param name="topics/cmd_vel" value="/$(arg tracker_frame)/cmd_vel" />
  </node>


  <!-- State estimators. -->
  <node name="position_velocity_yaw_state_estimator"
        pkg="crazyflie_state_estimator"
        type="position_velocity_yaw_state_estimator_node"
        output="screen">

    <param name="x_dim" value="$(arg lqr_x_dim)" />
    <param name="time_step" value="$(arg estimator_dt)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/robot" value="$(arg tracker_frame)" />

    <param name="topics/state" value="$(arg position_velocity_yaw_state_topic)" />
  </node>

  <node name="position_velocity_state_estimator"
        pkg="crazyflie_state_estimator"
        type="position_velocity_state_estimator_node"
        output="screen">

    <param name="x_dim" value="$(arg tracker_x_dim)" />
    <param name="time_step" value="$(arg estimator_dt)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/robot" value="$(arg tracker_frame)" />

    <param name="topics/state" value="$(arg position_velocity_state_topic)" />
  </node>

  <!-- LQR controller. -->
  <node name="lqr_controller"
        pkg="crazyflie_lqr"
        type="position_velocity_yaw_state_lift_lqr_node"
        output="screen">

    <param name="x_dim" value="$(arg lqr_x_dim)" />
    <param name="u_dim" value="$(arg lqr_u_dim)" />

    <param name="K_file" value="$(arg lqr_K_file)" />
    <param name="u_ref_file" value="$(arg lqr_u_ref_file)" />

    <param name="topics/reference" value="$(arg reference_state_topic)" />
    <param name="topics/state" value="$(arg position_velocity_yaw_state_topic)" />
    <param name="topics/control" value="$(arg lqr_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
  </node>

  <!-- Merger node. -->
  <node name="merger"
        pkg="crazyflie_control_merger"
        type="no_yaw_merger_node"
        output="screen">

    <param name="mode" value="$(arg merger_mode)" />

    <param name="topics/control" value="$(arg lqr_control_topic)" />
    <param name="topics/prioritized_control" value="$(arg optimal_control_topic)" />
    <param name="topics/merged" value="$(arg merged_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
    <param name="topics/reference" value="$(arg reference_state_topic)" />
  </node>

  <!-- Value function server node. -->
  <node name="neural_value_server"
        pkg="neural_tracker"
        type="neural_value_server_node.py"
        output="screen">

    <rosparam param="network_files" subst_value="True">$(arg network_files)</rosparam>

    <param name="srv/optimal_control" value="$(arg optimal_control_name)" />
    <param name="srv/tracking_bound" value="$(arg tracking_bound_name)" />
    <param name="srv/switching_tracking_bound" value="$(arg switching_bound_name)" />
    <param name="srv/guaranteed_switching_time" value="$(arg switching_time_name)" />
    <param name="srv/guaranteed_switching_distance" value="$(arg switching_distance_name)" />
    <param name="srv/priority" value="$(arg priority_name)" />
    <param name="srv/max_planner_speed" value="$(arg max_planner_speed_name)" />
    <param name="srv/best_possible_time" value="$(arg best_time_name)" />
  </node>

  <!-- Tracker, meta planner, and sensor nodes. -->
  <node name="tracker"
        pkg="meta_planner"
        type="tracker_node"
        output="screen">

    <param name="control/time_step" value="$(arg tracker_dt)" />
    <param name="control/dim" value="$(arg tracker_u_dim)" />
    <param name="state/dim" value="$(arg tracker_x_dim)" />
    <rosparam param="state/lower" subst_value="True">$(arg state_lower_bound)</rosparam>
    <rosparam param="state/upper" subst_value="True">$(arg state_upper_bound)</rosparam>

    <param name="srv/optimal_control" value="$(arg optimal_control_name)" />
    <param name="srv/priority" value="$(arg priority_name)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/tracker" value="$(arg tracker_frame)" />
    <param name="frames/planner" value="$(arg planner_frame)" />

    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
    <param name="topics/reference" value="$(arg reference_state_topic)" />
    <param name="topics/controller_id" value="$(arg controller_id_topic)" />
    <param name="topics/state" value="$(arg position_velocity_state_topic)" />
    <param name="topics/control" value="$(arg optimal_control_topic)" />
  </node>

  <node name="trajectory_interpreter"
        pkg="meta_planner"
        type="trajectory_interpreter_node"
        output="screen">

    <param name="max_runtime" value="$(arg max_meta_runtime)" />

    <param name="control/time_step" value="$(arg tracker_dt)" />
    <param name="control/dim" value="$(arg tracker_u_dim)" />
    <param name="state/dim" value="$(arg tracker_x_dim)" />
    <rosparam param="state/lower" subst_value="True">$(arg state_lower_bound)</rosparam>
    <rosparam param="state/upper" subst_value="True">$(arg state_upper_bound)</rosparam>

    <param name="srv/tracking_bound" value="$(arg tracking_bound_name)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/tracker" value="$(arg tracker_frame)" />
    <param name="frames/planner" value="$(arg planner_frame)" />

    <param name="topics/state" value="$(arg position_velocity_state_topic)" />
    <param name="topics/traj" value="$(arg traj_topic)" />
    <param name="topics/reference" value="$(arg reference_state_topic)" />
    <param name="topics/controller_id" value="$(arg controller_id_topic)" />
    <param name="topics/request_traj" value="$(arg request_traj_topic)" />
    <param name="topics/trigger_replan" value="$(arg trigger_replan_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
    <param name="topics/vis/traj" value="$(arg traj_vis_topic)" />
    <param name="topics/vis/tracking_bound" value="$(arg bound_vis_topic)" />
  </node>

  <node name="meta_planner"
        pkg="meta_planner"
        type="meta_planner_node"
        output="screen">

    <param name="random/seed" value="$(arg random_seed)" />
    <param name="max_runtime" value="$(arg max_meta_runtime)" />
    <param name="max_connection_radius" value="$(arg max_meta_connection_radius)" />

    <param name="control/dim" value="$(arg tracker_u_dim)" />
    <rosparam param="control/lower" subst_value="True">$(arg control_lower_bound)</rosparam>
    <rosparam param="control/upper" subst_value="True">$(arg control_upper_bound)</rosparam>

    <param name="state/dim" value="$(arg tracker_x_dim)" />
    <rosparam param="state/lower" subst_value="True">$(arg state_lower_bound)</rosparam>
    <rosparam param="state/upper" subst_value="True">$(arg state_upper_bound)</rosparam>

    <param name="planners/num_values" value="$(arg num_values)" />

    <param name="goal/x" value="$(arg goal_x)" />
    <param name="goal/y" value="$(arg goal_y)" />
    <param name="goal/z" value="$(arg goal_z)" />

    <param name="srv/tracking_bound" value="$(arg tracking_bound_name)" />
    <param name="srv/best_time" value="$(arg best_time_name)" />
    <param name="srv/switching_time" value="$(arg switching_time_name)" />
    <param name="srv/switching_distance" value="$(arg switching_distance_name)" />
    <param name="srv/switching_bound" value="$(arg switching_bound_name)" />

    <param name="topics/sensor" value="$(arg sensor_topic)" />
    <param name="topics/vis/known_environment" value="$(arg known_env_vis_topic)" />
    <param name="topics/traj" value="$(arg traj_topic)" />
    <param name="topics/state" value="$(arg position_velocity_state_topic)" />
    <param name="topics/request_traj" value="$(arg request_traj_topic)" />
    <param name="topics/trigger_replan" value="$(arg trigger_replan_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
  </node>

  <node name="sensor"
        pkg="meta_planner"
        type="sensor_node"
        output="screen">

    <param name="sensor/sensor_radius" value="$(arg sensor_radius)" />
    <param name="sensor/num_obstacles" value="$(arg num_obstacles)" />
    <param name="sensor/min_obstacle_radius" value="$(arg min_obstacle_radius)" />
    <param name="sensor/max_obstacle_radius" value="$(arg max_obstacle_radius)" />
    <param name="sensor/time_step" value="$(arg sensor_dt)" />

    <param name="control/dim" value="$(arg tracker_u_dim)" />
    <param name="state/dim" value="$(arg tracker_x_dim)" />
    <rosparam param="state/lower" subst_value="True">$(arg state_lower_bound)</rosparam>
    <rosparam param="state/upper" subst_value="True">$(arg state_upper_bound)</rosparam>

    <param name="random/seed" value="$(arg random_seed)" />

    <param name="srv/switching_bound" value="$(arg switching_bound_name)" />

    <param name="topics/sensor" value="$(arg sensor_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
    <param name="topics/vis/sensor_radius" value="$(arg sensor_radius_vis_topic)" />
    <param name="topics/vis/true_environment" value="$(arg true_env_vis_topic)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/tracker" value="$(arg tracker_frame)" />
  </node>
</launch>
